### DISCUSSION TITLE
Cache not working, what am I missing or doing it wrong?

### DISCUSSION THREAD

**Discussion by bairrada97** [vercel/community#575] [Help]
Hey, I have a next.js app where im setting up cache control on api route.

But it keeps hitting `x-vercel-cache: MISS`

```
export default async function handler(req, res) {
  const query = req.query;

  const response = await fetch('external-api-endpoint');
  const data = await response.json();
  res.setHeader("Cache-Control", `s-maxage=86000`);
  res.status(200).json(data);
}

```

I attached a screenshot of my network tab on response headers.
https://gyazo.com/72a697b8f65b7b85e19250f091bfa55e

**Comment by swarnava** [comment]
Hi there,

Have you tried configuring it using `vercel.json` ?

```
{
    "headers": [
        {
            "source": "/(.*)",
            "headers" : [
                {
                    "key" : "Cache-Control",
                    "value" : "public, max-age=31536000, immutable"
                }
            ]
        }
    ]
}
```

https://vercel.com/support/articles/how-to-configure-the-cache-control-response-header-in-vercel-projects

**Reply by bairrada97** [reply]
I haven't try it but I my use case its specific. I cant apply logic if I write my cache-control on `vercel.json`.

I have **2** cases.

**one is a page that has parameter Id e.g `/post/[id]`**.

I use `getServerSideProps` and inside I do a fetch call to external api and based on the response I want to cache for a day or for **1 week**. if 100 users go to /post/1 will see the page content not changing for 1 day and if they go to /post/2 it will cache for 1 week for example.

**and my other case is what Its described here on my question**.

inside my `post/[id]` I have a button that triggers an api route and I want to cache that api route for `s-maxage=86000`.

Both cases works inconsistently. 

if I test my api route I sometimes get this (see screenshot below) , looks like I only get `x-vercel-cache: HIT` only when I empty the cache of my browser. but I do not see `s-maxage=86000` on the response headers  instead I see `cache-control: public, max-age=0, must-revalidate.` Also if I go to other laptop or my smartphone  the cache returns `x-vercel-cache: MISS` again.

![Screenshot 2022-04-26 at 15 08 26](https://user-images.githubusercontent.com/26314199/165306907-720e18f7-15fc-4a86-9b4f-1bbe5144c417.png)

EDIT: forgot to mention that after I clear browser cache  and I go into the situation as described on the screenshot, so when cache apparently HITS, It will always works correctly, so I can go back to homepage and go into /post/[id] and click on the button and I will see instant result.  I can even open another tab or even go to icognito mode and will works perfect but works inconsistently as if I change laptop the cache its not returned
